### ISSUE TITLE
Wemo Light Switch Long Press

### ISSUE THREAD

**Issue by dlamkins** [pywemo/pywemo#61]
I could be wrong, but based on my understanding of the eventservice.xml file hosted by the WeMo lightswitches, it looks like the WeMo Light Switches now support the long press event via UPnP. My switches are running firmware version "WeMo_WW_2.00.10937.PVT-OWRT-LS" which I got on Jan 25, 2017. Here is my eventservice.xml contents:

[http://pastebin.com/zJgesnAc](http://pastebin.com/zJgesnAc)

I've highlighted the two sections regarding long presses. The first looks to be for simulating a long press on the WeMo and the second appears to be an event that can be subscribed to. Here are those sections pulled out:

**Action (In Action List)**

```
<action>
<name>SimulateLongPress</name>
<argumentList>
<argument>
<name>LongPress</name>
<direction>out</direction>
<relatedStateVariable>LongPress</relatedStateVariable>
</argument>
</argumentList>
</action>
```

**Event (In Service State Table)**

```
<stateVariable sendEvents="yes">
<name>LongPress</name>
<dataType>string</dataType>
<defaultValue>0</defaultValue>
</stateVariable>
```

Is it possible that this feature could get added?  I would really like to eventually be able to take advantage of this feature in Home-Assistant.

Thanks!

**Comment by pavoni** [comment]
Pull requests welcome. 

Afraid I don't have a Light Switch myself - so not much I can do.

**Comment by TaylorTWBrown** [comment]
If I donate a switch, can you build it? :D

**Comment by pavoni** [comment]
Afraid I'm based in the UK - and wemo don't support / supply the bridges or light switches here.

Don't know if @jaharkes would be open to help.

**Comment by jwkicklighter** [comment]
@pavoni I'm taking a look at this, but having trouble following how this library works. In `subscribe.py`, I see nothing that limits the type of events that are listened to. In theory, if the Wemo Switch is sending long press events shouldn't the logger catch that the event was received?

**Comment by pavoni** [comment]
@jwkicklighter it's easy to miss - take a look here:-
https://github.com/pavoni/pywemo/blob/master/pywemo/subscribe.py#L113

You can see the subscription to the `basicevent` service - and the commented out subscription to the `insightevent` service. The current code doesn't handle multiple subscriptions for a single device - hence commenting out the second subscription.

The HA wrapper refreshes the insight switch - so we get the state changes without the subscription (for now).

At some point I'll have another look at subscribing to multiple services. The subscriptions need to be renewed - so it's more complex than it first looks.

Other devices probably have similar subscription urls. easiest way to find out is to load them up - in an interactive python session - and then you can take a look. The services are dynamically created from data returned from the device.

I've never looked at the bridge devices (since I don't have one) - but from what I recall from when it was added - subscriptions didn't work for them - so it may be you can drop the basicevents service and replace it with something else...

Once you have the events you can either use the data in the event - or just force a refresh of the device state.

Let me know what you discover.

**Comment by jwkicklighter** [comment]
@pavoni Ah, seems more complex than what I first pictured. Thanks for the explanation though. I'll try to take another look soon, would love to get this added to the library (and eventually Home Assistant).

**Comment by SaunterinRoo** [comment]
Hello there,

I was able to find this discussion via the Home Assistant community forums. I was hoping to add to this, I recently added the new Wemo Dimmer wall switch to my setup. Of course being a brand new piece of hardware there is not yet support for it. Since this sort of development is way above my head, I was hoping to just start a discussion here about it, to see if that would be a possibility in the future to add support for as well as the long press functionality.

Thank you so much!

**Comment by pavoni** [comment]
@sircoolkeen The wemo api is basically undocumented - so the only way to get support is for a developer who has one of the devices to do some digging.

The good news is that the basic framework that talks to the devices and creates the apis it claims to support is already there - so it's not that difficult to add something (I got started trying to add support for the Wemo Maker which half worked). If the device is discovered OK - than you can inspect the device attributes via Python..

So perhaps have a go - may be easier than you expect...

**Comment by pavoni** [comment]
BTW - if the project this was forked from adds support it should be pretty easy to port it across (https://github.com/iancmcc/ouimeaux/issues/156)

**Comment by jwkicklighter** [comment]
@sircoolkeen You should also search for an issue related to the dimmer and/or make a new issue just to make sure that's documented.

@pavoni As far as getting the long press to work, I haven't had much luck trying to figure out the library and reverse engineer actually detecting the long press. Tried sniffing around with Wireshark, but didn't really get anywhere.

**Comment by SaunterinRoo** [comment]
@jwkicklighter Yup, I created an issue regarding adding support for the dimmer. After exhaustive searches I did find this.. eclipse/smarthome#3632

I'm not sure if this project could be ported across? 

Thanks, 
Keenan

**Comment by ikifar2012** [comment]
Any progress on adding long press to the Wemo’s pywemo?

**Comment by pavoni** [comment]
It basically needs a developer who has one to do this, even with an example to copy - you have to have the device to test the code. 

If you fancy having a go @ikifar2012 a PR would be very welcome.

**Comment by ikifar2012** [comment]
I have many WeMo devices but I am not that experienced in coding but I will be willing to help in any way I can

**Comment by jwkicklighter** [comment]
I had a look though and tried to do some monitoring on my network, but unfortunately I couldn't figure out how the library works well enough to make any progress 🙁

(I'm also not really a Python dev, which I'm sure is most of my problem).

**Comment by sqldiablo** [comment]
Hi, all. I'm taking code ownership of the WeMo platform on Home Assistant, and therefore, I'm working more on pywemo as well. I've picked up a series of the devices for debugging, so I'd be willing to look into this.

Right now, I'm working on an overhaul of the rediscovery process in pywemo, which will take me another week or two due to the holidays and complexity with how rediscovery works. After that, I'd be happy to look into this as well.

**Comment by pavoni** [comment]
@sqldiablo Have given you access to the repo, great to have some help!

**Comment by sqldiablo** [comment]
Thanks, @pavoni. Happy to help. 🙂

**Comment by TaylorTWBrown** [comment]
This is awesome! Thanks for the update.

**Comment by sqldiablo** [comment]
@pavoni I don't have a direct way to contact you, so I'm doing it on here, as it's semi-related to previous conversation. Would you be willing to set me as a maintainer for this module on pypi? My username on there is the same as here.

**Comment by pavoni** [comment]
@sqldiablo have added you. 

Thanks again for your interest in helping maintain & improve this.

**Comment by jwkicklighter** [comment]
@sqldiablo I was unable to ever make any progress on this, but if you need any testing done feel free to ping me!

**Comment by ikifar2012** [comment]
Has anyone made any progress on this? I really think it must be possible since IFTTT detects a long press

**Comment by jwkicklighter** [comment]
IFTTT is much different since it relies on the Wemo servers for its information. The local network traffic probably has the data available, but the ability for a cloud service to work isn't an indication that it's possible to do with a library such as this.

**Comment by esev** [comment]
FWIW on my [Light Switch](https://www.belkin.com/us/p/P-WLS040/) (`<firmwareVersion>WeMo_WW_2.00.11451.PVT-OWRT-LIGHTV2</firmwareVersion>`), I also see the same in /eventservice.xml

### /setup.xml
```
[snip]
      <service>
        <serviceType>urn:Belkin:service:basicevent:1</serviceType>
        <serviceId>urn:Belkin:serviceId:basicevent1</serviceId>
        <controlURL>/upnp/control/basicevent1</controlURL>
        <eventSubURL>/upnp/event/basicevent1</eventSubURL>
        <SCPDURL>/eventservice.xml</SCPDURL>
      </service>
[snip]
```

### /eventservice.xml
```
[snip]
  <serviceStateTable>
[snip]
    <stateVariable sendEvents="yes">
      <name>LongPress</name>
      <dataType>string</dataType>
      <defaultValue>0</defaultValue>
    </stateVariable>
[snip]
  </serviceStateTable>
  
  </scpd>
```

I tested receiving the LongPress event with the following code, but the light switch never sent a notification when the long press was activated. I double checked this with tcpdump also. It does send notifications for BinaryState changes (whenever the switch is turned on/off), but nothing for LongPress.
```
import logging
import pywemo
import time

def _cb(_device, _type, _params):
  logging.info('%s %s %s', _type, _params, _device.subscription_update(_type, _params))

logging.basicConfig(level=logging.DEBUG)
sub = pywemo.SubscriptionRegistry()
sub.start()

address = ''  # <-- Add IP address here
port = pywemo.ouimeaux_device.probe_wemo(address)
url = 'http://%s:%i/setup.xml' % (address, port)
device = pywemo.discovery.device_from_description(url, None)
print(device)

sub.on(device, None, _cb)
sub.register(device)
time.sleep(2)
try:
  sub.join()
except KeyboardInterrupt:
  sub.stop()
  sub.join()
```

I also tried implementing the STATEVAR header in the SUBSCRIBE request (`STATEVAR: BinaryState,LongPress`) and that didn't work either. The light switch never responded back with a ACCEPTED-STATEVAR header, nor did it start responding to LongPress state changes.

Since `basicevent` is the service corresponding to /eventservice.xml, I don't think implementing multiple subscriptions would solve this issue. Rather, IMHO, it appears that the light switch just doesn't send a notification for the LongPress state change.

I posted a question to the Wemo community forums, just in case any of their developers look there, but that's likely a long-shot.
https://community.wemo.com/s/question/0D52T000055hYWxSAM/longpress-state-variable-event-not-sent-for-upnp-subscription

**Comment by Spectre5** [comment]
I took a look at this too since it would be nice to use the long press to control any products within Home Assistant.  Unfortunately, unless Belkin changes how the long press works, I don't think it will be possible to subscribe to the long press.  As @esev already saw, there is no broadcast sent out for a long press.  I've verified that with wireshark as well.

The way long press is implemented is that the app sends a `StoreRules` action to the device with information that tells the device what to do when a long press is activated, i.e. what other wemo on the network to toggle, turn on, or turn off.  Then when the long press is used, that device sends an action directly to the controlled device --- one wemo sends a `SetBinaryState` to another wemo directly, without any generic broadcast that it is a long press event.  The `SimulateLongPress` would presumably be sent to a device to make it run that rule without needing to physically hold down the button.

The `<ruleDbBody>` body appears to be encrypted, so I'm not sure what the format for storing a rule is.  If we could figure that out, then we could create a rule and send it to a device, but it may only be able to send something to another wemo device on the network.  If that format was figured out and someone (not me!) was really adament about using the long press, then they could probably create a simple "listener" that could receive the `SetBinaryState` and then send that to Home Assistant as an MQTT message or similar.

**Comment by pavoni** [comment]
Given the comments above - I think this is too complex to consider for pywemo, so closing.

**Comment by esev** [comment]
@Spectre5 Real nice work!

Is the `<ruleDbBody>` perhaps a base64 zipped sqlite3 database?

```
import pywemo
device = pywemo.discovery.device_from_description('http://[snip-ip-port]/setup.xml', None)
device.rules.FetchRules()
{'ruleDbPath': 'http://[snip-ip-port]/rules.db', 'ruleDbVersion': '72', 'errorInfo': 'SUCCESS'}
```
```
curl http://[snip-ip-port]/rules.db --output rules.db
unzip rules.db
sqlite3 temppluginRules.db
.dump
```

```
SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> .dump
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE RULES ( RuleID PRIMARY KEY, Name TEXT NOT NULL, Type TEXT NOT NULL, RuleOrder INTEGER, StartDate TEXT, EndDate TEXT, State TEXT, Sync INTEGER);
INSERT INTO RULES VALUES('501','Kitchen Side Light Long Press Rule','Long Press',2,'12201982','07301982','1','NOSYNC');
CREATE TABLE RULEDEVICES(RuleDevicePK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , DeviceID TEXT , GroupID INTEGER , DayID INTEGER , StartTime INTEGER, RuleDuration INTEGER , StartAction REAL, EndAction REAL, SensorDuration INTEGER , Type INTEGER , Value INTEGER , Level INTEGER, ZBCapabilityStart TEXT, ZBCapabilityEnd TEXT,OnModeOffset INTEGER, OffModeOffset INTEGER, CountdownTime INTEGER, EndTime INTEGER);
INSERT INTO RULEDEVICES VALUES(21,501,'uuid:Lightswitch-3_0-[snip-serial-number]',0,-1,60,86340,0.0,-1.0,-1,-1,-1,-1,'','',-1,-1,-1,86400);
CREATE TABLE DEVICECOMBINATION(DeviceCombinationPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , SensorID TEXT , SensorGroupID INTEGER , DeviceID TEXT , DeviceGroupID INTEGER );
CREATE TABLE GROUPDEVICES(GroupDevicePK INTEGER PRIMARY KEY AUTOINCREMENT, GroupID INTEGER , DeviceID TEXT );
CREATE TABLE LOCATIONINFO(LocationPk INTEGER PRIMARY KEY AUTOINCREMENT, cityName TEXT, countryName TEXT, latitude TEXT, longitude TEXT, countryCode TEXT, region TEXT );
CREATE TABLE BLOCKEDRULES(Primarykey INTEGER PRIMARY KEY AUTOINCREMENT, ruleId TEXT );
CREATE TABLE RULESNOTIFYMESSAGE(RuleID INTEGER PRIMARY KEY AUTOINCREMENT, NotifyRuleID INTEGER , Message TEXT , Frequency INTEGER );
CREATE TABLE SENSORNOTIFICATION(SensorNotificationPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , NotifyRuleID INTEGER , NotificationMessage TEXT , NotificationDuration INTEGER );
CREATE TABLE TARGETDEVICES(TargetDevicesPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER, DeviceID TEXT, DeviceIndex INTEGER );
INSERT INTO TARGETDEVICES VALUES(26,501,'uuid:Dimmer-1_0-[snip-serial-number]',0);
INSERT INTO TARGETDEVICES VALUES(27,501,'uuid:Lightswitch-3_0-[snip-serial-number]',1);
INSERT INTO TARGETDEVICES VALUES(28,501,'uuid:Lightswitch-2_0-[snip-serial-number]',2);
CREATE TABLE android_metadata (locale TEXT);
INSERT INTO android_metadata VALUES('en_US');
DELETE FROM sqlite_sequence;
INSERT INTO sqlite_sequence VALUES('RULEDEVICES',21);
INSERT INTO sqlite_sequence VALUES('TARGETDEVICES',28);
COMMIT;
```

The long press rule I had stored in the device triggers three other devices to turn off.

**Comment by pavoni** [comment]
Wow - some amazing detective work going on here!

**Comment by esev** [comment]
I'm unable to capture the communication between the App and the Wemo device. I think all the communication from my App is flowing through the Wemo cloud service. 

@Spectre5 How were you able to get the app to communicate directly with the device? Would you be able to post an example of the `StoreRules` request with the `<ruleDbBody>`?

I see that my wemo is sending out SSDP discovery messages on its own. I suspect it is doing this to find the devices listed in the `TARGETDEVICES` table. I have my own SSDP responder that can answer to these messages and respond with the address/port of the pywemo subscription web server. And I can confirm that the wemo device then tries to fetch /setup.xml from the subscription web server. I further suspect that if I was able to update the rules database, that I could insert my own entry into the `TARGETDEVICES` table. But I can't figure out the right format for that `StoreRules` message.

I think with an example of the `StoreRules` message containing the `<ruleDbBody>` that I'd be able to get the long press event to happen on the pywemo subscription web server.

**Comment by Spectre5** [comment]
Hey, great work too!  You're right that most communication now goes through the cloud.  What I did to capture the communication is install Android 8 on a virtual machine and then installed an old APK (version 1.24 or older as the cloud change happened in version 1.25).  That older app version still works and then you can see the local messages!

I agree that the wemo is probably trying to discover devices itself to lookup the `TARGETDEVICES`, makes sense if it intends to send messages itself.

I went back to that pcap file and I've actually found that the app sent the exact same `StoreRules` to all 5 wemo devices on my network.  So it apparently wants to keep the rules in sync between the devices, or at least the long press rule.  Interestingly, I only make the long press rule control 1 device, so I'm not sure why it would bother to send the rule to the 3 three devices that are not involved.  Perhaps something in the database needs to be kept in sync?  The data it sends in the store rules also must then indicate which device the rule is for and which device is receiving the rule...because again, the exact same data was sent to each device.

First the app fetched the rules for all 5 devices, then it sent the `StoreRules` action to the target device.  When it got the `StoreRulesResponse` from that device indicating it was successful, it sent the `StoreRules` to the other 4 devices.  Finally, those 4 devices responded with a successful response as well.

I was going to try decoding that `ruleDbBody` when I have a few minutes, but I'm happy to have you do it if you have the time.  Since I'm not really sure if there is anything potentially sensitive in the data, I'd rather not post it here though, even though I can't really imagine anything sensitive being in it :)  Let me send it to you via e-mail.

**Comment by Spectre5** [comment]
ok, couldn't resist not looking at it closer right now.  Looks like you hit the nail on the head with the sqlite3 suggestion.  Here is what the app sent to the Wemo:

    <?xml version="1.0" encoding="utf-8"?>
    <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
       <s:Body>
          <u:StoreRules xmlns:u="urn:Belkin:service:rules:1">
             <ruleDbVersion>1</ruleDbVersion>
             <processDb>1</processDb>
             <ruleDbBody>&lt;![CDATA[LONG-DATA-STRING]]&gt;</ruleDbBody>
          </u:StoreRules>
       </s:Body>
    </s:Envelope>

I took out the `LONG-DATA-STRING` text and wrote it to a file using

    import base64
    with open('output.zip', 'wb') as fout:
        fout.write(base64.b64decode('LONG-DATA-STRING'))

Then I unzipped that file and there was a `temppluginRules.db` file that is readable by sqlite3:

    >>> unzip output.zip
    >>> sqlite3 temppluginRules.db 
    SQLite version 3.31.1 2020-01-27 19:55:54
    Enter ".help" for usage hints.
    sqlite> .dump
    PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE RULES ( RuleID PRIMARY KEY, Name TEXT NOT NULL, Type TEXT NOT NULL, RuleOrder INTEGER, StartDate TEXT, EndDate TEXT, State TEXT, Sync INTEGER);
    INSERT INTO RULES VALUES('501','Master Long Press Rule','Long Press',2,'12201982','07301982','1','NOSYNC');
    CREATE TABLE RULEDEVICES(RuleDevicePK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , DeviceID TEXT , GroupID INTEGER , DayID INTEGER , StartTime INTEGER, RuleDuration INTEGER , StartAction REAL, EndAction REAL, SensorDuration INTEGER , Type INTEGER , Value INTEGER , Level INTEGER, ZBCapabilityStart TEXT, ZBCapabilityEnd TEXT,OnModeOffset INTEGER, OffModeOffset INTEGER, CountdownTime INTEGER, EndTime INTEGER);
    INSERT INTO RULEDEVICES VALUES(1,501,'uuid:Lightswitch-1_0-[snip-serial-number]',0,-1,60,86340,2.0,-1.0,-1,-1,-1,-1,'','',-1,-1,-1,86400);
    CREATE TABLE DEVICECOMBINATION(DeviceCombinationPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , SensorID TEXT , SensorGroupID INTEGER , DeviceID TEXT , DeviceGroupID INTEGER );
    CREATE TABLE GROUPDEVICES(GroupDevicePK INTEGER PRIMARY KEY AUTOINCREMENT, GroupID INTEGER , DeviceID TEXT );
    CREATE TABLE LOCATIONINFO(LocationPk INTEGER PRIMARY KEY AUTOINCREMENT, cityName TEXT, countryName TEXT, latitude TEXT, longitude TEXT, countryCode TEXT, region TEXT );
    CREATE TABLE BLOCKEDRULES(Primarykey INTEGER PRIMARY KEY AUTOINCREMENT, ruleId TEXT );
    CREATE TABLE RULESNOTIFYMESSAGE(RuleID INTEGER PRIMARY KEY AUTOINCREMENT, NotifyRuleID INTEGER , Message TEXT , Frequency INTEGER );
    CREATE TABLE SENSORNOTIFICATION(SensorNotificationPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER , NotifyRuleID INTEGER , NotificationMessage TEXT , NotificationDuration INTEGER );
    CREATE TABLE TARGETDEVICES(TargetDevicesPK INTEGER PRIMARY KEY AUTOINCREMENT, RuleID INTEGER, DeviceID TEXT, DeviceIndex INTEGER );
    INSERT INTO TARGETDEVICES VALUES(1,501,'uuid:Socket-1_0-[snip-serial-number]',0);
    DELETE FROM sqlite_sequence;
    INSERT INTO sqlite_sequence VALUES('RULEDEVICES',1);
    INSERT INTO sqlite_sequence VALUES('TARGETDEVICES',1);
    COMMIT;

The fact that it fetches all rules on all devices and sent this to all devices makes me think (assume) that the indices for rules and such need to be unique on the devices?  Or maybe it just sends this to all devices and then devices that are not impacted by the new rules just ignore this file and don't bother updating their internal database.

All 5 devices responded with the same response:

    <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body>
    <u:StoreRulesResponse xmlns:u="urn:Belkin:service:rules:1">
    <errorInfo>Storing of rules DB Successful!
    </errorInfo>
    </u:StoreRulesResponse>
    </s:Body> </s:Envelope>

**Comment by esev** [comment]
That's what I was missing. Thank you!

I was trying to use `device.rules.StoreRules(ruleDbVersion=89, processDb=1, ruleDbBody='<![CDATA['+body+']]>')` which always resulted in `{'errorInfo': 'Storing of rules DB failed!!\n'}`

This seems to work `device.rules.StoreRules(ruleDbVersion=89, processDb=1, ruleDbBody='&lt;![CDATA['+body+']]&gt;')`. At least it returns `{'errorInfo': 'Storing of rules DB Successful!\n'}`

**Comment by Spectre5** [comment]
Very nice!  Making some real progress here!

**Comment by esev** [comment]
I have something working.  I added a virtual device to the `temppluginsRules.db` file.
```
INSERT INTO TARGETDEVICES VALUES(44,503,'uuid:Socket-1_0-VirtualDevice', 2);
```
Then zipped, base64'ed and updated the file with `device.rules.StoreRules`. I set `ruleDbVersion` to one greater than the value returned from `device.rules.FetchRules`

I also modified pywemo a bit to support some additional http verbs and added a `respond_to_discovery` method to respond to M-SEARCH requests. See https://github.com/esev/pywemo/commit/6e67613075427496ec871e20f321154672f834cb (still very much a work in progress).

Then I put it all together with something like this
```
import logging
import pywemo

logging.basicConfig(level=logging.DEBUG)
sub = pywemo.SubscriptionRegistry()
sub.start()

pywemo.ssdp.respond_to_discovery(('<my_ip_address>', 8989))
```

In the logs, I see `DEBUG:pywemo.subscribe:SUBSCRIBE /upnp/event/basicevent1` periodically as the wemo maintains a subscription to the virtual device within pywemo. Then, when I hold the button on the wemo to trigger a long-press, I see this `DEBUG:pywemo.subscribe:POST /upnp/control/basicevent1`, which I believe is the wemo sending my virtual device a command to turn on/off.

**Comment by Spectre5** [comment]
Really cool, great work

**Comment by esev** [comment]
This is still very much a work in progress, however I have a start on some code to automatically update the rules and make the long press trigger a subscription callback. I've added a `add_long_press_virtual_device` method to the `Device` base class that takes care of everything. See https://github.com/esev/pywemo/commit/34f08c9f66f0a724e9c8460bee989cb7359de0cf

I'm using it like this:

```
import logging
import pywemo

WEMO_IP = '' # Fill this in

def _cb(_device, _type, _params):
    logging.info('%s %s %s', _type, _params, _device.subscription_update(_type, _params))
    logging.info('State: %s', _device._state)

logging.basicConfig(level=logging.DEBUG)
sub = pywemo.SubscriptionRegistry()
sub.start()

device = pywemo.discovery.device_from_description(pywemo.discovery.setup_url_for_address(WEMO_IP, 0), None)
device.add_long_press_virtual_device() # Have the Long Press trigger a subscription callback
sub.register(device)
sub.on(device, 'LongPress', _cb)

pywemo.ssdp.respond_to_discovery(sub.port)
```

**Comment by Spectre5** [comment]
I haven't had a chance to test it yet, but I've taken a look at the code.  Certainly looks promising!  Two comments:

1.  We should add the ability to delete the long press rule too - either the long press created by pwemo or a long press rule that already exists (from the app).  That way a user can clear the long press rules and then add this virtual one easily.  Probably already on your todo list.
2. Just curios, how did you find out that the `ruleDbVersion` need to increment by 1 each time?

**Comment by esev** [comment]
1. Good plan! I added the ability to remove and generally interact with the database in a python friendly way in #190. Those helpers should allow us to create other types of rules in the future too (i.e. schedule rules).

2. I've been looking at the latest Android app and found this there.

**Comment by Spectre5** [comment]
For 2, how did you find this from the latest Android app?

**Comment by esev** [comment]
> For 2, how did you find this from the latest Android app?

There are some javascript resources within the app that give some hints. The version I'm looking at uses the cloud service, but it does the same, incrementing the ruleDbVersion. It might be okay to set it to 0.

One thing I've noticed is that any rules that I add through #190 get overwritten by the cloud service if the rules are updated in the App. I'm not sure there is any way around this.

**Comment by esev** [comment]
For what it's worth, I have this working in home-assistant now too.
https://github.com/home-assistant/core/compare/dev...esev:wemo_long_press

```
Event 0 fired 12:31 PM:
{
    "event_type": "wemo.long_press",
    "data": {
        "entity_id": "switch.kitchen_side_light"
    },
    "origin": "LOCAL",
    "time_fired": "2020-12-22T20:31:17.595027+00:00",
    "context": {
        "id": "7d4dff54e87eb2e846050219c4a63b52",
        "parent_id": null,
        "user_id": null
    }
}
```

**Comment by esev** [comment]
I think there is some opportunity to notice that the rules were updated via the App and then re-insert the pywemo virtual device. Since the wemo devices SUBSCRIBE to the SubscriptionRegistry when they are configured to communicate with the virtual device, it'd be possible to keep track of which devices have stopped subscribing. The rules for the devices that stop subscribing could then be updated again.

That's probably more work than I'm wanting to take on right now though. I'm happy enough to just never configure the rules via the App again. Just noting that this could be a possible solution for users that want to use both the App and pywemo.

**Comment by esev** [comment]
Just a quick update. I've been running with these two patches for the past few days. Everything seems to be working reliably, including the new long press device triggers in HA.

- pywemo: https://github.com/pavoni/pywemo/compare/master...esev:long_press
- home-assistant: https://github.com/home-assistant/core/compare/dev...esev:wemo_long_press

My plan is break up the large pywemo change into smaller PRs that are easier for review:

1. Add Python classes for interacting with the rules database (#190).
2. Support a virtual device to receive long press events (#192)
3. Inform Wemo devices of the pywemo virtual Wemo device (#193)
4. Add long press support to LightSwitch and Dimmer devices (#194).

1 is the largest change, so I'll wait on the review for that one before sending the other 3 PRs.

**Comment by Spectre5** [comment]
Great to hear that you're been running this with success.  I haven't had the time to look closely at the PR yet myself, but I will try to within the next 1 or 2 days.  I agree with your thought process on breaking it up into a few separate PR though since the database portion is entirely new to pywemo.

**Comment by dereitz** [comment]
This is exciting stuff!  Looking forward to being able to use it (when included)!!

**Comment by esev** [comment]
Thanks for the reviews, for testing, and for all the investigation @Spectre5 

Give me a heads-up after the next release happens and I'll start pulling this into HomeAssistant.
